docs(grid): Add Cell Selection documentation and revamp Selection articles (#2211)
telerik/blazor-docs: components/grid/selection/rows.md

---
title: Row Selection
page_title: Grid - Rows Selection
description: Learn how to select row in Blazor Grid component. Explore the selected rows. Discover row selection bevahior when combined with other Grid features. Try the practical sample code for row selection.
slug: components/grid/selection/rows
tags: telerik,blazor,grid,selection,rows
previous_url: /components/grid/selection/single,/components/grid/selection/multiple
position: 3
---

# Row Selection

The Grid component offers support for single or multiple row selection. You can select a row with mouse click and through checkbox column. You can access the collection of selected rows, use this collection and manipulate it. You can follow and respond to the event of selection.

In this article:

* [Rows Selection Options](#rows-selection-options)
	* [Click-Only Selection](#click-selection)

> 	* [Checkbox Selection](#checkbox-selection)

Reviewer: Merge these three sections into one with a single example. We already have a checkbox column example in the Columns section.

    * [Selection in Template](#selection-in-template)
    * [Selection and Row Drag and Drop](#selection-and-row-drag-and-drop)

## Rows Selection Options

By default you can [click](#click-selection) on the row to select it. Additionally, you can have a [checkbox selection](#checkbox-selection). The checkbox selection allows the user to select the row through a dedicated on the row checkbox. You can combine click and checkbox selection or use only one selection option.

You can use row selection with both [selection modes]({%slug components/grid/selection/overview%}#selection-mode)—single and multiple.

### Click Selection

By default, users can select rows by clicking anywhere in the row, except on the command buttons.

To select multiple rows, hold down the `Ctrl` or `Shift` key to extend the selection:
* Press and hold `Ctrl` and click the desired rows to select or deselect them.
* Click on the starting row in a range of rows that you want to select, press and hold `Shift`, and click on the last row in the range. The first selected row is the starting point of the range and the last selected row is the end of the selection.

> If you release the `Ctrl` or the `Shift` keys and click to start new multiple selection, the previously selected rows will be deselected.

Reviewer: We describe keyboard navigation in the online demos. You can link to it and remove this content.

>caption Click selection and single selection mode

````CSHTML
Click on one row to select it.

<TelerikGrid Data=@GridData

>              SelectionMode="@GridSelectionMode.Single"

Reviewer: It is not a real-world scenario to enable selection and not define `SelectedItems`.

             Pageable="true">
    <GridColumns>
        <GridColumn Field=@nameof(Employee.Name) />

>         <GridColumn Field=@nameof(Employee.Team) Title="Team" />

Reviewer: Use quotes around parameter values.

                {
                    EmployeeId = i,
                    Name = "Employee " + i.ToString(),
                    Team = "Team " + i % 3
                });
        }
    }

    public class Employee
    {
        public int EmployeeId { get; set; }
        public string Name { get; set; }
        public string Team { get; set; }
    }
}
````

### Checkbox Selection

You can also render a checkbox column that allows users to select and deselect rows. To use checkbox selection, add a [`GridCheckboxColumn`]({%slug components/grid/columns/checkbox%}) in the `GridColumns` collection of the Grid. The `GridCheckboxColumn` provides [additional configuration settings related to selection]({%slug components/grid/columns/checkbox%}#parameters).

To deselect the row, click its checkbox again.

>caption Checkbox selection and multiple selection mode

````CSHTML
@if (CheckBoxOnlySelection)
{
    <div>You can <strong>click only the checkbox itself</strong> to select or deselect the row. You cannot click the entire row.</div>
}
else
{
    <div>You can <strong>click both the checkbox or the row</strong> to select or deselect the row.</div>
}
<br/>

> <TelerikCheckBox @bind-Value="@CheckBoxOnlySelection"
>                  Id="@CheckboxId"/>
> <label for="@CheckboxId">Toggle checkbox only selection</label>

Reviewer: You don't need Id in this case.

```suggestion
<label><TelerikCheckBox @bind-Value="@CheckBoxOnlySelection" /> Toggle checkbox only selection</label>
```

* You can use the `SelectedItems` collection in two-way binding. You can predefine the selected item for your users through the two-way binding of the `SelectedItems` property. The collection will be updated by the Grid when the selection changes.

### Selected Rows When Data Changes

When the Grid `Data` collection changes, the `SelectedItems` collection has the following behavior:

* When you update a selected item in the Grid, you have to make the same update in the `SelectedItems` collection through the Grid [editing events]({%slug components/grid/editing/overview%}).
* When you delete a selected item in the Grid, it will automatically delete from the `SelectedItems` collection. If you are using one-way binding for the `SelectedItems` collection and the [`SelectedItemsChanged` event](#selecteditemschanged), when you delete a selected item, the event fires. When you delete all selected items, the `SelectedItemsChanged` event fires with an empty collection.
* When you create an item in the Grid, and you want to select it at the same time, use the Grid [editing events]({%slug components/grid/editing/overview%}).

> ### Selected Rows Equals Comparison

Reviewer: This whole section is 99% the same in the Row and Cell pages. Consider moving it to the Overview page.

Author: This will change the idea of the Overview page:
- to contain only the most important information in a summary
- to redirect for more information to the row and cell articles
I prefer to leave it as it is, so when the user is reading the row/cell article, they will have the full information in one place

## SelectedItemsChanged

You can respond to the user action of selecting a new row through the `SelectedItemsChanged` event. The `SelectedItemsChanged` event receives a collection of the Grid data model. It may have no items in it. It may have only one member (the last selected item) when the `SelectionMode` is `Single`.

>caption Single selection with one-way binding for SelectedItems and using the SelectedItemsChanged event

````CSHTML
<TelerikGrid Data=@GridData
             SelectionMode="@GridSelectionMode.Single"
             SelectedItemsChanged="@((IEnumerable<Employee> employeeList) => OnSelect(employeeList))"

>              SelectedItems="@SelectedEmployees"

Reviewer: Define parameters before their respective Changed handler

             Pageable="true"
             Height="400px">
    <GridColumns>
        <GridCheckboxColumn />
        <GridColumn Field=@nameof(Employee.Name) />
        <GridColumn Field=@nameof(Employee.Team) Title="Team" />
    </GridColumns>
</TelerikGrid>

@if (SelectedEmployees != null)
{
    <ul>
        @foreach (Employee employee in SelectedEmployees)
        {
            <li>
                @employee.Name
            </li>
        }
    </ul>
}

@code {

>     private List<Employee> GridData { get; set; }

Reviewer: ```suggestion
    private List<Employee> GridData { get; set; } = new();
```

    private IEnumerable<Employee> SelectedEmployees { get; set; } = Enumerable.Empty<Employee>();

>     private Employee SelectedEmployee { get; set; }

Reviewer: Objects should be nullable or instantiated immediately. If they are nullable, you may need to adjust other code in the example. The ultimate goal is to not have any compiler warnings.

```suggestion
    private Employee? SelectedEmployee { get; set; }
```

    protected override void OnInitialized()
    {
        GridData = new List<Employee>();

>         for (int i = 0; i < 15; i++)

Reviewer: Do not use 0 as an ID value. In general, do not use the default value of a type as an actual ID value.

        {
            GridData.Add(new Employee()
                {
                    EmployeeId = i,

>                     Name = "Employee " + i.ToString(),

Reviewer: Shorter:
```suggestion
                    Name = $"Employee {i}",
```

    private IEnumerable<Employee> SelectedEmployees { get; set; } = Enumerable.Empty<Employee>();
    private Employee SelectedEmployee { get; set; }
    protected override void OnInitialized()
    {
        GridData = new List<Employee>();
        for (int i = 0; i < 15; i++)
        {
            GridData.Add(new Employee()
                {
                    EmployeeId = i,
                    Name = "Employee " + i.ToString(),
                    Team = "Team " + i % 3
                });
        }
    }

    protected void OnSelect(IEnumerable<Employee> employees)
    {
        SelectedEmployee = employees.FirstOrDefault();
        // update the collection so that the grid can highlight the correct item
        // when two-way binding is used this happens automatically,
        // but the framework does not allow two-way binding and the event at the same time
        SelectedEmployees = new List<Employee> { SelectedEmployee };
    }

    public class Employee
    {
        public int EmployeeId { get; set; }
        public string Name { get; set; }
        public string Team { get; set; }
    }
}
````

### SelectedItemsChanged and Asynchronous Operations

> If you want to execute asynchronous operations, such as loading data on demand, when the user selects a row, this should be handled in the [`OnRowClick`]({%slug grid-events%}#onrowclick) or [`OnRowDoubleClick`]({%slug grid-events%}#onrowdoubleclick) events rather than in the [`SelectedItemsChanged`]({%slug grid-events%}#selecteditemschanged).

Reviewer 2: ```suggestion
To execute asynchronous operations, such as loading data on demand, when the user selects a row, handle these operations in the [`OnRowClick`]({%slug grid-events%}#onrowclick) or [`OnRowDoubleClick`]({%slug grid-events%}#onrowdoubleclick) events rather than in the [`SelectedItemsChanged`]({%slug grid-events%}#selecteditemschanged).
```

> ## Row Selection and Other Grid Features

Reviewer 2: Add a sentence that explains what you present in this section and how the information can help the developer.

### Selection with Editing Modes

> #### InCell Edit Mode

Reviewer 2: * `#### InCell Edit Mode` and `#### Inline and Popup Edit Modes` should be list items and not H4. The content in the `#### Inline and Popup Edit Modes` section is limited and doesn't justify using an entire section.
* Add a sentence below the `### Selection with Editing Modes` heading and use it to introduce the H4s that you will turn into list items in an unordered list.
